pkg/migration/generate.go

```diff
@@ -0,0 +1,165 @@
+)
+
+func (h *Handler) addStorkSpec(cluster *corev1.StorageCluster) error {
+	storkDeployment := &appsv1.Deployment{}
+	err := h.client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      storkDeploymentName,
+			Namespace: cluster.Namespace,
+		},
+		storkDeployment,
+	)
+	if errors.IsNotFound(err) {
+		cluster.Spec.Stork = &corev1.StorkSpec{
+			Enabled: false,
+		}
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	cluster.Spec.Stork = &corev1.StorkSpec{
+		Enabled: true,
+	}
+	return nil
+}
+
+func (h *Handler) addAutopilotSpec(cluster *corev1.StorageCluster) error {
+	autopilotDeployment := &appsv1.Deployment{}
+	err := h.client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      autopilotDeploymentName,
+			Namespace: cluster.Namespace,
+		},
+		autopilotDeployment,
+	)
+	if errors.IsNotFound(err) {
+		cluster.Spec.Autopilot = &corev1.AutopilotSpec{
+			Enabled: false,
+		}
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	cluster.Spec.Autopilot = &corev1.AutopilotSpec{
+		Enabled: true,
+	}
+	return nil
+}
+
+func (h *Handler) addPVCControllerSpec(cluster *corev1.StorageCluster) error {
+	pvcDeployment := &appsv1.Deployment{}
+	err := h.client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      pvcControllerDeploymentName,
+			Namespace: cluster.Namespace,
+		},
+		pvcDeployment,
+	)
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Explicitly enable pvc controller in kube system namespace as operator won't
+	// enable it by default if running in kube-system namespace
+	if cluster.Namespace == "kube-system" {
+		cluster.Annotations[pxutil.AnnotationPVCController] = "true"
+	}
+	return nil
+}
+
+func (h *Handler) addMonitoringSpec(cluster *corev1.StorageCluster) error {
+	// Check if metrics need to be exported
+	svcMonitorFound := true
+	svcMonitor := &monitoringv1.ServiceMonitor{}
+	err := h.client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      serviceMonitorName,
+			Namespace: cluster.Namespace,
+		},
+		svcMonitor,
+	)
+	if errors.IsNotFound(err) {
+		svcMonitorFound = false
+	} else if err != nil {
+		return err
+	}
+	cluster.Spec.Monitoring.Prometheus = &corev1.PrometheusSpec{
+		ExportMetrics: svcMonitorFound,
+	}
+
+	// Check for alert manager
+	alertManagerFound := true
+	alertManager := &monitoringv1.Alertmanager{}
+	err = h.client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      alertManagerName,
+			Namespace: cluster.Namespace,
+		},
+		alertManager,
+	)
+	if errors.IsNotFound(err) {
+		alertManagerFound = false
+	} else if err != nil {
+		return err
+	}
+	cluster.Spec.Monitoring.Prometheus.AlertManager = &corev1.AlertManagerSpec{
+		Enabled: alertManagerFound,
+	}
+
+	if !cluster.Spec.Monitoring.Prometheus.ExportMetrics &&
+		!cluster.Spec.Monitoring.Prometheus.AlertManager.Enabled {
+		return nil
+	}
+
+	// Check for prometheus operator
+	prometheusOpDeployment := &appsv1.Deployment{}
```

Comment:
this code block is used in several places, it can be put in a function doesDeploymentExist() (bool, error)

Reply:
Done.
Instead of a boolean, just returning the actual deployment as in one place we need the deployment object.